das/daser.go

```diff
@@ -68,195 +72,82 @@ func (d *DASer) Start(context.Context) error {
+	if atomic.CompareAndSwapInt32(&d.closed, 0, 1) {
+		d.cancel()
 
-// catchUpManager manages catch-up jobs, performing them one at a time, exiting
-// only once context is canceled and storing latest DASed checkpoint to disk.
-func (d *DASer) catchUpManager(ctx context.Context, checkpoint int64) {
-	defer func() {
-		// store latest DASed checkpoint to disk here to ensure that if DASer is not yet
-		// fully caught up to network head, it will resume DASing from this checkpoint
-		// up to current network head
-		// TODO @renaynay: Implement Share Cache #180 to ensure no duplicate DASing over same
-		//  header
-		if err := storeCheckpoint(ctx, d.cstore, checkpoint); err != nil {
-			log.Errorw("storing checkpoint to disk", "height", checkpoint, "err", err)
+		// shutdown the sampler
+		if err := d.sampler.finished(ctx); err != nil {
+			if errors.Is(err, context.DeadlineExceeded) {
+				return fmt.Errorf("DASer force quit: %w", err)
+			}
+			return err
 		}
-		log.Infow("stored checkpoint to disk", "checkpoint", checkpoint)
-		// signal that catch-up routine finished
-		d.catchUpDn <- struct{}{}
-	}()
 
-	for {
+		// wait for subscriber to quit
 		select {
+		case <-d.subscriberDone:
 		case <-ctx.Done():
-			return
-		case job := <-d.jobsCh:
-			// record details of incoming job
-			d.recordJobDetails(job)
-			// perform catchUp routine
-			height, err := d.catchUp(ctx, job)
-			// store the height of the last successfully sampled header
-			checkpoint = height
-			// exit routine if a catch-up job was unsuccessful
-			if err != nil {
-				// record error
-				d.state.catchUpLk.Lock()
-				d.state.catchUp.Error = err
-				d.state.catchUpLk.Unlock()
-
-				log.Errorw("catch-up routine failed", "attempted range: from", job.from, "to", job.to)
-				log.Warn("DASer CATCH-UP SAMPLING ROUTINE WILL BE STOPPED. IN ORDER TO CONTINUE SAMPLING, " +
-					"RE-START THE NODE")
-				return
-			}
+			return ctx.Err()
 		}
 	}
+	return nil
 }
 
-// catchUp starts a sampling routine for headers starting at the next header
-// after the `from` height and exits the loop once `to` is reached. (from:to]
-func (d *DASer) catchUp(ctx context.Context, job *catchUpJob) (int64, error) {
-	log.Infow("sampling past headers", "from", job.from, "to", job.to)
-
-	// start sampling from height at checkpoint+1 since the
-	// checkpoint height is DASed by broader sample routine
-	for height := job.from + 1; height <= job.to; height++ {
-		h, err := d.getter.GetByHeight(ctx, uint64(height))
-		if err != nil {
-			if err == context.Canceled {
-				// report previous height as the last successfully sampled height and
-				// error as nil since the routine was ordered to stop
-				return height - 1, nil
-			}
-
-			log.Errorw("failed to get next header", "height", height, "err", err)
-			// report previous height as the last successfully sampled height
-			return height - 1, err
-		}
+func (d *DASer) fetch(ctx context.Context, height uint64) error {
```

Comment:
Let's call it `sampleHeight`. Understand where is fethc coming from but it's a bit off

Reply:
ty, renamed for clarity